File: parser.py
"""将 Obsidian Canvas JSON 解析为 Card 对象。"""

import json
import re
from pathlib import Path

from models import Card, GREEN_COLOR, DEFAULT_DECK, CARD_META_PATTERN


def parse_canvas(canvas_path: Path) -> tuple[list[Card], list[str], list[Card]]:
    """解析 .canvas 文件，返回 (cards, warnings, deletions)。

    cards：绿色 + 有 --- 的节点，待导出。
    warnings：绿色但无 --- 的节点（跳过）。
    deletions：非绿色但有 ID 的节点（待删除）。
    """
    with open(canvas_path, "r", encoding="utf-8") as f:
        data = json.load(f)

    nodes = data.get("nodes", [])
    edges = data.get("edges", [])

    node_map = {n["id"]: n for n in nodes}

    # 出边映射：source_id → [target_ids]
    outgoing: dict[str, list[str]] = {}
    for edge in edges:
        src = edge["fromNode"]
        outgoing.setdefault(src, []).append(edge["toNode"])

    groups = [n for n in nodes if n.get("type") == "group"]

    cards: list[Card] = []
    warnings: list[str] = []
    deletions: list[Card] = []

    for node in nodes:
        if node.get("type") != "text":
            continue

        is_green = node.get("color") == GREEN_COLOR
        text = node.get("text", "")
        meta = _extract_meta(text)
        clean_text = _strip_meta(text)

        if is_green:
            front, back = _split_qa(clean_text)
            if front is None:
                preview = clean_text[:20]
                warnings.append(f'SKIP: 节点 "{preview}" — 绿色但无 --- 分隔')
                continue

            tags, front = _extract_tags(front)
            back_tags, back = _extract_tags(back)
            tags.extend(back_tags)

            # 出边内容追加到答案末尾（按 y 坐标排序）
            target_ids = outgoing.get(node["id"], [])
            if target_ids:
                targets = [node_map[tid] for tid in target_ids if tid in node_map]
                targets.sort(key=lambda n: n.get("y", 0))
                for target in targets:
                    back = _append_target(back, target)

            deck = _find_deck(node, groups)

            cards.append(Card(
                node_id=node["id"],
                front=front.strip(),
                back=back.strip(),
                deck=deck,
                tags=tags,
                anki_id=meta.get("id"),
            ))

        elif not is_green and meta.get("id"):
            # 非绿色 + 有 ID → 待删除
            preview = clean_text[:20]
            deletions.append(Card(
                node_id=node["id"],
                front=preview,
                back="",
                anki_id=meta["id"],
            ))

    return cards, warnings, deletions


def _split_qa(text: str) -> tuple[str | None, str | None]:
    """在第一个裸 --- 处分割文本，返回 (正面, 背面) 或 (None, None)。"""
    lines = text.split("\n")
    for i, line in enumerate(lines):
        if line.strip() == "---":
            # 前一行以 | 开头则是表格，跳过
            if i > 0 and lines[i - 1].strip().startswith("|"):
                continue
            # 后一行以 | 开头则是表格，跳过
            if i + 1 < len(lines) and lines[i + 1].strip().startswith("|"):
                continue
            return "\n".join(lines[:i]), "\n".join(lines[i + 1:])
    return None, None


def _extract_meta(text: str) -> dict:
    """从文本提取 <!--card:{...}--> 元数据。"""
    match = re.search(CARD_META_PATTERN, text)
    if match:
        try:
            return json.loads(match.group(1))
        except json.JSONDecodeError:
            return {}
    return {}


def _strip_meta(text: str) -> str:
    """从文本中移除 <!--card:{...}-->。"""
    return re.sub(r"<!--card:.*?-->", "", text).strip()


def _extract_tags(text: str) -> tuple[list[str], str]:
    """提取 #标签，返回 (标签列表, 去标签后的文本)。"""
    tags = re.findall(r"#([\w\u4e00-\u9fff]+)", text)
    cleaned = re.sub(r"#[\w\u4e00-\u9fff]+", "", text)
    return tags, cleaned


def _append_target(back: str, target: dict) -> str:
    """将出边目标节点的内容追加到答案末尾。"""
    t = target.get("type")
    if t == "text":
        return back + "\n\n" + target.get("text", "")
    elif t == "file":
        filepath = target.get("file", "")
        ext = filepath.rsplit(".", 1)[-1].lower() if "." in filepath else ""
        if ext in ("png", "jpg", "jpeg", "gif", "svg", "webp"):
            return back + f"\n\n![[{filepath}]]"
        else:
            name = filepath.rsplit("/", 1)[-1].rsplit(".", 1)[0]
            return back + f"\n\n[[{name}]]"
    return back


def _find_deck(node: dict, groups: list[dict]) -> str:
    """找到包含该节点的 Group，返回其 label；未找到则返回 DEFAULT_DECK。"""
    nx, ny = node["x"], node["y"]
    nw, nh = node["width"], node["height"]
    for g in groups:
        gx, gy = g["x"], g["y"]
        gw, gh = g["width"], g["height"]
        if nx >= gx and ny >= gy and nx + nw <= gx + gw and ny + nh <= gy + gh:
            return g.get("label", DEFAULT_DECK)
    return DEFAULT_DECK

File: plugin/src/parser.ts
import { Card, CARD_META_RE, CARD_META_GLOBAL_RE, DEFAULT_DECK } from "./models";

// ── types for raw Canvas JSON ──────────────────────────────────────────────
interface RawNode {
  id: string;
  type: "text" | "file" | "group" | string;
  text?: string;
  file?: string;
  label?: string;
  color?: string;
  x: number;
  y: number;
  width: number;
  height: number;
}

interface RawEdge {
  id: string;
  fromNode: string;
  toNode: string;
}

interface CanvasData {
  nodes: RawNode[];
  edges: RawEdge[];
}

// ── public result type ─────────────────────────────────────────────────────
export interface ParseResult {
  cards: Card[];
  warnings: string[];
  deletions: Card[];
}

// ── helpers ────────────────────────────────────────────────────────────────

/**
 * Split text on the FIRST bare `---` separator (not inside a code fence).
 * Returns [front, back] or null if no bare separator exists.
 */
export function splitQA(text: string): [string, string] | null {
  const lines = text.split("\n");
  let inFence = false;
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    if (line.trimEnd().match(/^```/)) {
      inFence = !inFence;
    }
    if (!inFence && line.trim() === "---") {
      const front = lines.slice(0, i).join("\n");
      const back = lines.slice(i + 1).join("\n");
      return [front, back];
    }
  }
  return null;
}

/** Parse `<!--card:{JSON}-->` and return the id field, or null. */
export function extractMeta(text: string): number | null {
  const m = CARD_META_RE.exec(text);
  if (!m) return null;
  try {
    const obj = JSON.parse(m[1]);
    return typeof obj.id === "number" ? obj.id : null;
  } catch {
    return null;
  }
}

/** Remove all `<!--card:...-->` occurrences from text. */
export function stripMeta(text: string): string {
  return text.replace(CARD_META_GLOBAL_RE, "").trim();
}

/** Extract #hashtags (ASCII + CJK), return tags array and cleaned text. */
export function extractTags(text: string): { tags: string[]; text: string } {
  const TAG_RE = /#([\w\u4e00-\u9fff]+)/g;
  const tags: string[] = [];
  const cleaned = text.replace(TAG_RE, (_match, tag) => {
    tags.push(tag);
    return "";
  }).replace(/\s{2,}/g, " ").trim();
  return { tags, text: cleaned };
}

/** Build the append string for a linked target node. */
export function appendTarget(node: RawNode): string {
  if (node.type === "file" && node.file) {
    const lower = node.file.toLowerCase();
    const isMedia = /\.(png|jpg|jpeg|gif|svg|webp|mp4|pdf)$/.test(lower);
    return isMedia ? `![[${node.file}]]` : `[[${node.file}]]`;
  }
  if (node.type === "text" && node.text) {
    return stripMeta(node.text);
  }
  return "";
}

/** Return the label of the smallest group that contains (cx, cy), or null. */
export function findDeck(
  cx: number,
  cy: number,
  groups: RawNode[]
): string | null {
  let best: RawNode | null = null;
  let bestArea = Infinity;
  for (const g of groups) {
    if (
      cx >= g.x &&
      cx <= g.x + g.width &&
      cy >= g.y &&
      cy <= g.y + g.height
    ) {
      const area = g.width * g.height;
      if (area < bestArea) {
        bestArea = area;
        best = g;
      }
    }
  }
  return best?.label ?? null;
}

// ── main export ────────────────────────────────────────────────────────────

export function parseCanvas(json: string, exportColor: string, deleteGroupLabel: string): ParseResult {
  const data: CanvasData = JSON.parse(json);
  const nodes: RawNode[] = data.nodes ?? [];
  const edges: RawEdge[] = data.edges ?? [];

  const cards: Card[] = [];
  const warnings: string[] = [];
  const deletions: Card[] = [];

  // Build lookup maps
  const nodeById = new Map<string, RawNode>(nodes.map((n) => [n.id, n]));
  const groups = nodes.filter((n) => n.type === "group");

  // Identify DELETE group (by label match)
  const deleteGroups = groups.filter(
    (g) => g.label?.trim() === deleteGroupLabel.trim()
  );

  /** Returns true if node center is inside any DELETE group */
  function isInDeleteGroup(node: RawNode): boolean {
    const cx = node.x + node.width / 2;
    const cy = node.y + node.height / 2;
    return deleteGroups.some(
      (g) => cx >= g.x && cx <= g.x + g.width && cy >= g.y && cy <= g.y + g.height
    );
  }

  // Outgoing edges per node
  const edgesFrom = new Map<string, string[]>();
  for (const e of edges) {
    if (!edgesFrom.has(e.fromNode)) edgesFrom.set(e.fromNode, []);
    edgesFrom.get(e.fromNode)!.push(e.toNode);
  }

  for (const node of nodes) {
    if (node.type !== "text" || !node.text) continue;

    const rawText = node.text;
    const ankiId = extractMeta(rawText);

    // DELETE group branch: color-independent, only checks ankiId
    if (isInDeleteGroup(node)) {
      if (ankiId !== null) {
        deletions.push({
          nodeId: node.id,
          front: "",
          back: "",
          deck: DEFAULT_DECK,
          tags: [],
          ankiId,
        });
      } else {
        warnings.push(`Node ${node.id}: in DELETE group but has no ankiId — skipped`);
      }
      continue;
    }

    // Normal export branch: color must match
    const isMatch = node.color === exportColor;
    if (!isMatch) continue;

    // Matching node — must have a QA separator
    const clean = stripMeta(rawText);
    const split = splitQA(clean);
    if (!split) {
      warnings.push(`Node ${node.id}: no Q/A separator found`);
      continue;
    }

    let [rawFront, rawBack] = split;

    // Append edge targets to back, sorted by y position
    const targetIds = edgesFrom.get(node.id) ?? [];
    const targets = targetIds
      .map((id) => nodeById.get(id))
      .filter((n): n is RawNode => !!n)
      .sort((a, b) => a.y - b.y);

    for (const t of targets) {
      const appended = appendTarget(t);
      if (appended) rawBack = rawBack + "\n" + appended;
    }

    // Extract tags from front and back
    const frontResult = extractTags(rawFront.trim());
    const backResult = extractTags(rawBack.trim());
    const tags = [...new Set([...frontResult.tags, ...backResult.tags])];

    // Determine deck via group containment (center point of node)
    const cx = node.x + node.width / 2;
    const cy = node.y + node.height / 2;
    const deck = findDeck(cx, cy, groups) ?? DEFAULT_DECK;

    cards.push({
      nodeId: node.id,
      front: frontResult.text,
      back: backResult.text,
      deck,
      tags,
      ankiId,
    });
  }

  return { cards, warnings, deletions };
}
